fix(#1701): Events page displays year for events scheduled to start or end at a future year

alvindera97 opened the pull request:
## What
- Addition of event year text to events page (for events scheduled with any date **not** in the current year.
- Replaces https://github.com/python/pythondotorg/pull/1714.

## Closes
- Fixes https://github.com/python/pythondotorg/issues/1701.

alvindera97 committed:
Add Methods To Check If Event Starts And Ends This Year

alvindera97 committed:
Set Up Templates For Querying Start And End Years

Passed variables to the time_tag template [time_tag.html] that checks if an event was scheduled to start or end with the current year.

alvindera97 committed:
Insert New Test Data And Update test_views.py

More events are created to test particular scenarios of events especially events set to start or end at a future year.

alvindera97 committed:
Time Tag Now Shows Year For Events With Details Not Within The Current Year

The time tag now displays the year when an event will occur. This is only for events that have been scheduled to start or end in at a future year. The accompanying functional tests have also been included.

alvindera97 committed:
Move All Test Data To Functional Test

All test data concerning the provision of data to serve the functional tests have been moved to the functional test.
As it improves readability.
All other test data at test_views.py was reset
to accommodate for the reduction in number of test data instances.

alvindera97 committed:
Functional Test For Displaying Year Of Event For Future Events Now Implemented With Unit Tests

alvindera97 committed:
Remove Functional Test For Displaying Year Of Future Event

Since the current CI at the main branch does not support selenium [web driver] operations, the functional test which depends on selenium to run has been removed.

alvindera97 committed:
Handle Case When Call To Next Event Returns None

alvindera97 committed:
Fix Erroneous Addition To Dev Requirements

This error was introduced in 115af08e7ce529e2e9298d93a98aacf77bb007af

alvindera97 committed:
Merge remote-tracking branch 'upstream/main' into fix-issue-1701

JacobCoffee committed:
Merge branch 'main' into fix-issue-1701

JacobCoffee commented on events/tests/test_models.py:
```diff
@@ -186,3 +185,25 @@ def test_event_previous_event(self):
+    def test_scheduled_to_start_this_year_method(self):
+        now = seconds_resolution(timezone.now())
+
+        occurring_time_dtstart = now + datetime.timedelta(days=3)
```
Wouldnt this fail if it were the last 2 days of the year?

JacobCoffee commented on events/tests/test_models.py:
```diff
@@ -186,3 +185,25 @@ def test_event_previous_event(self):
+    def test_scheduled_to_end_this_year_method(self):
+        now = seconds_resolution(timezone.now())
+
+        occurring_time_dtstart = now + datetime.timedelta(days=3)
```
same

alvindera97 replied:
Same comment here concerning 
```
occurring_time_dtstart = now + datetime.timedelta(days = 3)
``` 
with previous comment at https://github.com/python/pythondotorg/pull/2500/files#r1764230987

**If it's fine with everyone, I'll await feedback first before making any more changes to the pull request so I'd attempt resolving all concerns with as minimal back-and-forth as possible.**

Thank you for your time and best regards everyone.

alvindera97 commented on events/tests/test_models.py:
```diff
@@ -186,3 +185,25 @@ def test_event_previous_event(self):
+    def test_scheduled_to_start_this_year_method(self):
+        now = seconds_resolution(timezone.now())
+
+        occurring_time_dtstart = now + datetime.timedelta(days=3)
```
Hmmmm... Thinking about this test again. On the off chance that tests are run between the 29th of December and the 31st of December, this test will fail since the next year will be set to 3 days ahead...

So if tests are run on the 29th of December, `occurring_time_dtstart` will be set to 3 days from then which would be the 1st of January (the following year).

Since the date comparison is made against `datetime.datetime.now().year` at `Event.is_scheduled_to_start_this_year` and `Event.is_scheduled_to_end_this_year`, (https://github.com/alvindera97/pythondotorg/blob/fa6e0ced4827a1652aa2eef0a80db22b1a2c4fe2/events/models.py#L184-L189) having a timedelta of 3 days makes running tests within the last 3 days of the year flaky.

hugovk replied:
You can use something like https://github.com/spulec/freezegun to freeze time so the tests run reliably.

alvindera97 replied:
> You can use something like https://github.com/spulec/freezegun to freeze time so the tests run reliably.

Thanks, @hugovk, for suggesting [freezegun](https://github.com/spulec/freezegun) to handle the time-dependent tests. I have tested the updates locally, and I’m ready to push the changes to the current pull request. However, before proceeding, I wanted to get clarity on one remaining concern regarding the introduction of external dependencies into the project.

Specifically, I’m unsure of the current team policy regarding the addition of new dependencies. My concern revolves around potential risks that come with relying on third-party libraries, such as maintenance issues, security vulnerabilities, and compatibility with future versions of Python/Django.

In this case, while freezegun is a robust library for freezing time, I wonder if it’s necessary to introduce an entire dependency solely to address the time-delta requirements of the Event object in the tests. I'm considering whether there might be an alternative approach that avoids the need for an external library, such as directly mocking time or using built-in Python features to clamp dates.

I would appreciate your thoughts on whether the use of freezegun is acceptable for this situation, or if the team would prefer exploring another solution.

However, as an example, I thought of reducing the time-deltas. Here's a snippet **without freezegun**: 
```
    def test_scheduled_to_start_this_year_method(self):
        now = seconds_resolution(timezone.now())

        occurring_time_dtstart = now + datetime.timedelta(seconds=10)
        OccurringRule.objects.create(
            event=self.event,
            dt_start=occurring_time_dtstart,
            dt_end=occurring_time_dtstart + datetime.timedelta(days=3)
        )
        self.assertTrue(self.event.is_scheduled_to_start_this_year())

        OccurringRule.objects.get(event=self.event).delete()

        event_not_scheduled_to_start_this_year_occurring_time_dtstart = now + datetime.timedelta(days=365)
        OccurringRule.objects.create(
            event=self.event,
            dt_start=event_not_scheduled_to_start_this_year_occurring_time_dtstart,
            dt_end=event_not_scheduled_to_start_this_year_occurring_time_dtstart + datetime.timedelta(days=3)
        )

        self.assertFalse(self.event.is_scheduled_to_start_this_year())

    def test_scheduled_to_end_this_year_method(self):
        now = seconds_resolution(timezone.now())
        occurring_time_dtstart = now + datetime.timedelta(seconds=10)

        OccurringRule.objects.create(
            event=self.event,
            dt_start=occurring_time_dtstart,
            dt_end=occurring_time_dtstart
        )

        self.assertTrue(self.event.is_scheduled_to_end_this_year())

        OccurringRule.objects.get(event=self.event).delete()

        event_not_scheduled_to_end_this_year_occurring_time_dt_end = now + datetime.timedelta(days=365)

        OccurringRule.objects.create(
            event=self.event,
            dt_start=now,
            dt_end=event_not_scheduled_to_end_this_year_occurring_time_dt_end
        )

        self.assertFalse(self.event.is_scheduled_to_end_this_year())

```

Looking forward to your feedback 🙂

hugovk replied:
Thanks for your question, it's always good to keep these things in mind. My feeling is that it would be okay, as there are already a number of third-party dependencies, but I'll leave it to the maintainers to answer more concretely.

An alternative would be to use something like `unitest.mock` to fake the date returned by `datetime.timedelta`.

JacobCoffee replied:
Freezgun is fine, it should go into here https://github.com/python/pythondotorg/blob/7928f4fb6eb77f94623534a8caf854d0dbe61b67/dev-requirements.txt#L3 @alvindera97

alvindera97 replied:
Good day @hugovk and all,

I've opted to mock timezone (https://github.com/alvindera97/pythondotorg/blame/c4db8d42591f359eb899a5940ea25b71aebcf8a3/events/tests/test_models.py#L191-L247). While Django's import path for timezone seems stable, any future changes would require updating the mock path. Currently, I'm not aware of any deprecations or changes affecting timezone.

I'm open to further feedback, but in the absence of any, I'd like to thank you all again for your valuable input and prompt responses.

alvindera97 committed:
Refactor Tests For Relevant Year String Rendering At Events Page

- Updated the test methods `test_scheduled_to_start_this_year_method` and `test_scheduled_to_end_this_year_method` to better reflect event scheduling edge cases.
- Added assertions to verify when events are not scheduled to start or end within the current year.
- Utilize `unittest.mock` to clamp down datetime-sensitive tests.

alvindera97 committed:
Merge remote-tracking branch 'origin/main' into fix-issue-1701

alvindera97 committed:
Merge remote-tracking branch 'upstream/main' into fix-issue-1701

JacobCoffee committed:
Merge branch 'main' into fix-issue-1701

JacobCoffee reviewed (approved):
<img width="874" alt="image" src="https://github.com/user-attachments/assets/7cd5fae5-df83-4ad3-b8d5-8c19896dadb4">
lgtm
